### DISCUSSION TITLE
2.6.6 on ARM64

### DISCUSSION THREAD

**Discussion by nelseric** [rbenv/ruby-build#1946] [Build failures] [answered]
I've managed to install 2.7.2 and 3.0.0, but 2.6.6 fails with the following error

```
compiling closure.c
closure.c:264:14: error: implicit declaration of function 'ffi_prep_closure' is
      invalid in C99 [-Werror,-Wimplicit-function-declaration]
    result = ffi_prep_closure(pcl, cif, callback, (void *)self);
             ^
1 error generated.
```

**Comment by hsbt** [comment]
It's known issue. Ruby 2.6 didn't support M1 macOS yet. 

If you want to use Ruby 2.6 with M1 macOS, Can you file it to https://bugs.ruby-lang.org/? Thanks.

**Comment by nelseric** [comment]
Thanks for pointing me in the right direction.

**Comment by mrkn** [comment]
@nelseric I guess the error is only due to fiddle.  Probably, you can install Ruby 2.6 with specifying `--with-out-ext=fiddle` in the configure option.

**Reply by philicious** [reply]
This worked like a charm for me. Even for Ruby 2.5.1 (..ye dont ask)

**Comment by nelseric** [comment]
@mrkn Thanks, that worked

**Comment by rience** [comment]
I was also able to install Ruby 2.6.6 with this

```
RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC rbenv install 2.6.6
```

**Comment by santostiago** [comment]
I ended up doing this:

`export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include";   export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rbenv install 2.5.5`

**Reply by guizaols** [reply]
> I ended up doing this:
> 
> `export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"; export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rbenv install 2.5.5`

It worked for me as well! Thanks a lot!

**Comment by SalvatoreT** [comment]
@santostiago, that worked for me. Thank you!

**Comment by ak15** [comment]
> I ended up doing this:
> 
> `export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"; export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rbenv install 2.5.5`

This worked for me in My Apple M1 - Please try this if you are facing issue (I tried for ruby v 2.6.0) and don't need rosetta too


@santostiago

Once again thanks worked with 2.6.6

**Comment by HundredBillion** [comment]
THIS WORKED FOR ME
I spent 3hrs trying to get ruby 2.6.2 installed on my M1.

These 3 lines of code worked for me. Just make sure to change the version number to 2.6.2 (if that is what you need)

> `export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"; export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rbenv install 2.5.5`

**Comment by edmangimelli** [comment]
thank you so much @santostiago 😭

this actually solved my `asdf` install problem: (fish shell)
```
set -gx optflags "-Wno-error=implicit-function-declaration"
set -gx LDFLAGS "-L/opt/homebrew/opt/libffi/lib"
set -gx CPPFLAGS "-I/opt/homebrew/opt/libffi/include"
set -gx PKG_CONFIG_PATH "/opt/homebrew/opt/libffi/lib/pkgconfig"
asdf install ruby 2.6.6
```

**Comment by jujudellago** [comment]
> I ended up doing this:
> 
> `export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"; export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rbenv install 2.5.5`

more thanks for you @santostiago !!!  you saved my day !!

- oh edit... now that I got ruby compiled with rbenv or asdf successfully, now looking for a way to actually run rails....

endless ffi errors

`/Users/juju/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/ffi-1.9.18/lib/ffi/types.rb:69:in `find_type': unable to resolve type 'size_t' (TypeError)`

**Comment by olivierlacan** [comment]
@jujudellago Did you try to update the ffi gem? Or uninstall and reinstall it with the same flags shown above prefixed before you run bundle install? Bundler accepts build flags configuration it can store for specific gems. See: https://bundler.io/v1.16/bundle_config.html#BUILD-OPTIONS

**Comment by jujudellago** [comment]
Oh, so actually my error was just related to my project, I had a hardcoded version of ffi in my gemfile 

To figure this out I just made an empty rails app, saw everything was fine, and realized I had the wrong ffi in gemfile.lock

With other compatibility issues I ended up using ruby 2.7.2

**Comment by soheil** [comment]
`rvm install 2.6.6 --with-out-ext=fiddle` works great.

**Comment by olivierlacan** [comment]
I’m sorry the flag is `--with-out-ext`? 

Seriously? 
Not `--without-ext`, but `--with-out-ext`? 😳

Can we please fix that if it wasn’t a typo on your side? Seems bad.

**Comment by tf-rshearer** [comment]
`rvm install 2.6.6 --with-out-ext=fiddle` Worked for me on Big Sur 11.4 w/ m1. Thanks!

**Comment by michaelsmith-st** [comment]
RVM is much more fussy on the installation process (pgp keys, openSSL installation). And after all that, 
`rvm install 2.6.6 --with-out-ext=fiddle` **did not work** for me on Big Sur 11.4 w/ M1. 

The combination of `export` commands with **`rbenv install 2.6.6` above worked great**. Plus the dependency installation is soooo much smoother. 

I use RVM on my Intel Mac...but I'm inclined to switch to rbenv now. (My use case is simple, so I skipped the `rbenv-gem-rehash` and `gem install` parts). 
https://thoughtbot.com/blog/using-rbenv-to-manage-rubies-and-gems

**Comment by BGMP** [comment]
@michaelsmith-st Were you able to get Ruby 2.6.6 to install via RVM in the end? I haven't had any luck so far, and pretty much every answer pointing to its installation through the use of flags and so on ends up as a [broken] install for me.

**Comment by etipton** [comment, marked as answer]
> I was also able to install Ruby 2.6.6 with this
> 
> ```
> RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC rbenv install 2.6.6
> ```

Just wanted to chime in and confirm that this is safe to do... I didn't really understand what this would actually _do_ so I dug in a bit... wanted to be sure this wouldn't cause unexpected issues down the line*

To cut to the chase...

You can see [here](https://github.com/ruby/fiddle/blob/9be9eddf7a821e7ae45b8fd8ee1ecba78bacd4fe/ext/fiddle/closure.c#L260-L264) that by setting the USE_FFI_CLOSURE_ALLOC macro, the offending function (ffi_prep_closure) is never referenced.

And to confirm that the macro makes sense (the ffi_closure_alloc function is indeed available):

```
irb(main):001:0> require 'mkmf'
=> true
irb(main):002:0> pkg_config('libffi')
=> ["", "", "-lffi"]
irb(main):003:0> have_library('ffi')
checking for -lffi... yes
=> true
irb(main):004:0> have_func('ffi_closure_alloc', 'ffi.h')
checking for ffi_closure_alloc() in ffi.h... yes
=> true
```

*In hindsight, I guess that's what compilation is for in the first place, lol

**Reply by mislav** [reply]
Marking this as the answer because it expands on the original answer by providing great context, thank you!

FWIW, I was able to build Ruby 2.6.10 on an M1 machine without any extra compiler flags, but people's mileage may vary, and setting the `USE_FFI_CLOSURE_ALLOC` macro might still be a valid solution to compilation problems referencing `ffi_prep_closure`.

**Comment by yeukfei02** [comment]
encounter same problem in macbook m1

```
export optflags="-Wno-error=implicit-function-declaration"; export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"; export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"; export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig" | rvm install ruby-2.6.5
```

I tried this I am able to install ruby 2.6.5, not sure if this `rvm install 2.6.5 --with-out-ext=fiddle` also work in m1
just to share to let others know

**Comment by genezys** [comment]
Thanks for all the comments. Here is the exact commands I used to install 2.6.7 using rbenv on an M1 Pro:

```
export optflags="-Wno-error=implicit-function-declaration"
export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"
export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"
export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
rbenv install 2.6.7
```

I hope this multi-line version will be more readable to others. 
Based on previous comments, it should work for RVM too.

**Reply by cilim** [reply]
Be careful with this! I installed this 2 years ago, and in newer versions of Ruby (3.1.2) had a serious performance impact:

```ruby
require 'benchmark'

Benchmark.measure { (1..100000).inject(:*) || 1 }
```

@real=13.773456999999325

After removing these flags from .zshrc and reinstalling Ruby, it's fast again:

@real=4.0806959999990795

**Comment by rimian** [comment]
> Thanks for all the comments. Here is the exact commands I used to install 2.6.7 using rbenv on an M1 Pro:
> 
> ```
> export optflags="-Wno-error=implicit-function-declaration"
> export LDFLAGS="-L/opt/homebrew/opt/libffi/lib"
> export CPPFLAGS="-I/opt/homebrew/opt/libffi/include"
> export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig"
> rbenv install 2.6.7
> ```
> 
> I hope this multi-line version will be more readable to others. Based on previous comments, it should work for RVM too.

This worked with chruby and 2.6.5 for me. Thanks!
